### PR TITLE
Improvements

### LINKED ISSUE #19: UI improvements
Depends on https://github.com/ubiquity-os/ubiquity-os-plugin-installer/pull/13

What should be done:
1. The "select configuration" step should be removed. Always use the "production" config.
2. Show plugin readme on installing a new plugin (original [comment](https://github.com/ubiquity-os/ubiquity-os-plugin-installer/issues/15#issue-2646661367))
3. Show indicator if plugin is already installed (original [comment](https://github.com/ubiquity-os/ubiquity-os-plugin-installer/issues/16#issue-2647181893))
4. Load current config values in the editor (if config is present)
5. Add "Reset to defaults" button which resets plugin options to default values (original [comment](https://github.com/ubiquity-os/ubiquity-os-plugin-installer/issues/17#issue-2647183231))
6. Add github redirect to the same page ([example](https://github.com/ubiquity/work.ubq.fi/blob/5ad2731d99f3c03cdb52fe5f7122e412669ca0bd/src/home/rendering/render-github-login-button.ts#L37))

### PR DESCRIPTION
**Pull request by Keyrxng**
Resolves #19

Changes:
- readme display
- production config only
- reset to defaults

QA:

https://github.com/user-attachments/assets/20fb3e2c-994f-46be-8a1c-bb8e27e2678a

### CODE REVIEW

**Comment by Keyrxng** [comment]
This PR became quite long so I will address tests in another task

@rndquu @gentlementlegen if you'd like to take a look when you can, cheers.

**Comment by gentlementlegen** [comment]
Seems to work fine although I have a few remarks:
- I don't seem to have padding like in your video
image

- the left panel shifts a bit when hovering the "use default" button
- I am only seeing two orgs but I am in more orgs that that, how are they chosen?
- going back from the edit plugin menu makes the bar take the full width
image
- the "remove" button should maybe be disabled on plugins that are not installed

**Comment by Keyrxng** [comment]
> I am only seeing two orgs but I am in more orgs that that, how are they chosen?

https://docs.github.com/en/rest/orgs/orgs?apiVersion=2022-11-28#list-organizations-for-the-authenticated-user

```ts
public async getGitHubUserOrgs(): Promise<string[]> {
    const octokit = await this.getOctokit();
    const response = await octokit.rest.orgs.listForAuthenticatedUser();
    return response.data.map((org: { login: string }) => org.login);
  }
```

What orgs did you see?

I've considered other methods but the question is: What orgs do you want to see? 

- Orgs that you are an admin of or just owner of, collaborator of?
- Should the partner determine which roles can edit the config or do we enforce rules, if we enforce then what should they be?

I think a task should be created and it discussed there because it'll involve additional work and direction.

We sort of scrapped global config variables from `ubiquity-os-config.yml` but we could leverage that to enforce partner defined roles which can edit (at least via the UI). 

E.g: The org owner must use it after `kernel-setup` and we allow them to input roles that can edit and we write this to the config separate from `plugins:` and we read this on fetch and check against our authenticated user.

**Comment by Keyrxng** [comment]
> the left panel shifts a bit when hovering the "use default" button

I couldn't reproduce this pls lmk if it happens again and what you suggest as a fix/steps to repro, cheers

**Comment by zugdev** [comment]
We should only show "no orgs available" after we've looked at the request answer, it can be confusing in high latency environments. Show "loading", either as string or animation. Perhaps copy from [work.ubq.fi](https://github.com/ubiquity/work.ubq.fi/blob/5ad2731d99f3c03cdb52fe5f7122e412669ca0bd/static/style/style.css#L817C5-L819C41).

**Comment by zugdev** [comment]
1. Though I've selected an org there is no visual indicator of what org I am at, this can be confusing.

2. README fails to be opened for me. Also, no `console.error` log to help debug this.

image

**Comment by Keyrxng** [comment]
> README fails to be opened for me. Also, no console.error log to help debug this.

Any chance you wrote a log to find out as I cannot repro and I don't think it happened for mentlegen either or he'd have mentioned it?

I can add a console.error but would like to know the root cause as me adding a log won't resolve it, appreciate if you could.

> We should only show "no orgs available"

I thought this notification was enough but I have updated the title for while it's fetching also

image

File: static/main.ts
Comment (zugdev) [inline static/main.ts]: This is a duplicate cache set, you already cache manifests inside `fetchMarketplaceManifests()` on L41 of `fetch-manifest.ts`:

`localStorage.setItem("manifestCache", JSON.stringify(manifestCache));`

File: static/scripts/rendering/plugin-select.ts
Comment (zugdev) [inline static/scripts/rendering/plugin-select.ts]: You have `JSON.parse(localStorage.getItem("manifestCache") || "{}")` more than once, perhaps write a getter function.

**Comment by zugdev** [comment]
These icons are not intuitive, perhaps text is better.

image

File: static/index.html
Comment (zugdev) [inline static/index.html:29]: If there is still time we should move buttons outside of table. Tables are terrible for layouting and will definitely generate padding/overflow problems eventually. I think the config itself is fine within table as there is no easy better way to do it.

**Comment by Keyrxng** [comment]
> These icons are not intuitive, perhaps text is better.
> 
> image

They do have text when hovered and were part of 0x4007's original implementation I just tried to adopt a similar approach for reset to defaults but if you think just text would be better I can do that lmk.

image

**Comment by zugdev** [comment]
Some responsiveness issues, not sure if relevant as of this PR:

image

File: static/index.html
Comment (Keyrxng) [reply static/index.html:29]: Move them to where? I lack the imagination lmao. Perhaps the little sprinkles and sparkles can be handled as one PR once this is stable? https://github.com/ubiquity-os/ubiquity-os-plugin-installer/issues/14 will be the last change to the UI for for V1

**Comment by zugdev** [comment]
json is not vertically expanding for me:

https://github.com/user-attachments/assets/cc0b8c1c-e623-47bf-a8cf-0e0822d7197e

**Comment by zugdev** [comment]
Once a plugin config is added, a popup appears to push to GitHub. I think we either directly push (kinda dangerous), or offer a "shopping cart" where changes get added to your "cart" and partner reviews it all before pushing.

image

Also, if I push and go back the plugin won't appear as green/added until I refresh:

https://github.com/user-attachments/assets/3405fd4f-b263-4b24-8e39-8f86d29fccb5

**Comment by Keyrxng** [comment]
> json is not vertically expanding for me:
> 
>  d.mp4

1. That is relevant to https://github.com/ubiquity-os/ubiquity-os-plugin-installer/issues/14. If you don't have that plugin installed all of those fields render as `null` and specific handling has not been implemented yet.
2. You'll notice that it does expand on other plugins

> Once a plugin config is added, a popup appears to push to GitHub. I think we either directly push (kinda dangerous), or offer a "shopping cart" where changes get added to your "cart" and partner reviews it all before pushing.

If you ignore or close the notification you should be able to continue building other plugins continuously, so there is a shopping cart in effect but hidden. Think it's worth a little legend display what plugins are stored but not yet pushed?

> Some responsiveness issues, not sure if relevant as of this PR:

Not sure exactly what you are referring to, I assume the height mismatch? 

Some plugins render a < 250px modal, `convo-rewards` is the largest and is going to become much larger after https://github.com/ubiquity-os/ubiquity-os-plugin-installer/issues/14, so just introduce a basic scroll for the editor and capped it's size to the readme?

**Comment by zugdev** [comment]
> Any chance you wrote a log to find out as I cannot repro and I don't think it happened for mentlegen either or he'd have mentioned it?

Cleared cache and is now working smooth, perhaps it was on my end. If I can repro again I'll debug.

> I thought this notification was enough but I have updated the title for while it's fetching also

I think styling decisions are more opinion based comments than direct request for changes. Just let it be for now, unless agreed by team.

**Comment by Keyrxng** [comment]
> Also, if I push and go back the plugin won't appear as green/added until I refresh:

Resolved here [7bd56ac](https://github.com/ubiquity-os/ubiquity-os-plugin-installer/pull/25/commits/7bd56acec90ae8d1a3bf12f608094dc778ed08f9)

> > I thought this notification was enough but I have updated the title for while it's fetching also

> I think styling decisions are more opinion based comments than direct request for changes. Just let it be for now, unless agreed by team.

I implemented the title change alongside the notification but I'm going to leave the buttons alone unless told that it should just be text. 

As for button placement, I think they look good where they are but if I'm told to place them somewhere specific I can do that.

**Comment by Keyrxng** [comment]
> @Keyrxng, this task has been idle for a while. Please provide an update.

waiting for the team

**Comment by gentlementlegen** [comment]
> > I am only seeing two orgs but I am in more orgs that that, how are they chosen?
> 
> https://docs.github.com/en/rest/orgs/orgs?apiVersion=2022-11-28#list-organizations-for-the-authenticated-user
> 
> ```ts
> public async getGitHubUserOrgs(): Promise<string[]> {
>     const octokit = await this.getOctokit();
>     const response = await octokit.rest.orgs.listForAuthenticatedUser();
>     return response.data.map((org: { login: string }) => org.login);
>   }
> ```
> 
> What orgs did you see?
> 
> I've considered other methods but the question is: What orgs do you want to see?
> 
> * Orgs that you are an admin of or just owner of, collaborator of?
> * Should the partner determine which roles can edit the config or do we enforce rules, if we enforce then what should they be?
> 
> I think a task should be created and it discussed there because it'll involve additional work and direction.
> 
> We sort of scrapped global config variables from `ubiquity-os-config.yml` but we could leverage that to enforce partner defined roles which can edit (at least via the UI).
> 
> E.g: The org owner must use it after `kernel-setup` and we allow them to input roles that can edit and we write this to the config separate from `plugins:` and we read this on fetch and check against our authenticated user.

I have seen only 2 out of 4 orgs where I am an admin with the rights to create a configuration (`Meniole` and `ubiquity-os`). I should have seen `ubiquity-os-marketplace` most likely.

**Comment by Keyrxng** [comment]
> I have seen only 2 out of 4 orgs where I am an admin with the rights to create a configuration (Meniole and ubiquity-os). I should have seen ubiquity-os-marketplace most likely.

image

image

You are an `Owner` in both of those orgs @gentlementlegen so I'd expect marketplace to show up too but idk why it isn't. I don't share your permissions so I cannot debug this personally because any other org I use I'll be the creator of lmao.

What are your thoughts? 

- Marketplace org permission/access settings affecting the API res?
- wrong API method being used?

---

I did try `listPermissionsforUser` which lists `member`-`admin` etc but the gotcha is that an owner shares the same role as an admin so it's not straight forward to determine if they own it.

**Comment by Keyrxng** [comment]
>  I should have seen ubiquity-os-marketplace most likely.

So this UI should display all orgs that the user is an admin of then? Or simply all orgs in which you have `write` permissions?

If I know clearly what orgs should be displayed (defined by role or some other tangible fact) then I can make efforts to make it happen.

**Comment by zugdev** [comment]
Of course the ideal is to show precisely the ones he can edit. If that can't be done it's better to show all admin orgs than to miss some.

File: static/scripts/rendering/org-select.ts
Comment (zugdev) [inline static/scripts/rendering/org-select.ts]: redundant cache set here, handleOrgSelection already caches

**Review by zugdev — CHANGES_REQUESTED** [review]

**Comment by zugdev** [comment]
I can't add conversation rewards plugin with default config, I imagine it's because of the undefineds and nulls. An error message should indicate that.

[cant-add-conversation-rewards.webm](https://github.com/user-attachments/assets/ef7627af-062d-4f06-a0bf-cdbdd6e48d2a)

**Comment by zugdev** [comment]
When trying to add activity watcher, 5 errors pop up at the same time regarding the plugin URL being unavailable.

[watcher.webm](https://github.com/user-attachments/assets/b1adc185-6c62-4254-9758-75ad01867e30)

**Comment by zugdev** [comment]
When trying to add wallet, 7 success pop ups. I am thinking the pop ups are accumulating, will investigate.

image

**Comment by zugdev** [comment]
writeNewConfig executions are accumulating:

[accumulating-toaster.webm](https://github.com/user-attachments/assets/de6d1c8b-f95d-490d-bdf1-8c88254ae800)

It's outside of toaster, you can follow the experiment by adding `console.log("1");` inside of `writeNewConfig(...)` function declaration. Then go into a plugin add it, go back, add it go back, it follows an arithmetic progression each time you do that.

[accumulating-writeConfigs.webm](https://github.com/user-attachments/assets/4e449e78-8fb4-44ab-bf18-bb522a861e60)

File: static/scripts/rendering/config-editor.ts
Comment (zugdev) [inline static/scripts/rendering/config-editor.ts]: the problem is this addEventListener statement, if only you navigate a couple times and then execute an action it will stackup. you only need one event listener please check that. 

[eventListenerAdded.webm](https://github.com/user-attachments/assets/976db9d7-b09b-42c5-9663-72d10b858773)

**Comment by zugdev** [comment]
Remove is broken for me, nothing happens when I press it.

File: static/scripts/rendering/write-add-remove.ts
Comment (zugdev) [inline static/scripts/rendering/write-add-remove.ts]: typo: space between sentences here

File: static/scripts/rendering/write-add-remove.ts
Comment (zugdev) [inline static/scripts/rendering/write-add-remove.ts]: please use `else if (option === "remove")`, if new option is added someday can lead to error

File: static/scripts/rendering/write-add-remove.ts
Comment (zugdev) [inline static/scripts/rendering/write-add-remove.ts]: typo: space between sentences

File: static/scripts/rendering/config-editor.ts
Comment (zugdev) [inline static/scripts/rendering/config-editor.ts]: the function isn't being called because it's in a arrow function. please copy the syntax from the add button:

 remove.addEventListener("click", writeNewConfig.bind(null, renderer, "remove"));

File: static/scripts/rendering/config-editor.ts
Comment (zugdev) [inline static/scripts/rendering/config-editor.ts]: pressing reset to default is disabling the remove button when it shouldn't please mind that you should pass in plugin to `renderConfigEditor`

**Review by rndquu — CHANGES_REQUESTED** [review]
@Keyrxng 

1. Getting this error on trying to add the https://github.com/ubiquity-os-marketplace/text-conversation-rewards plugin:
Screenshot 2024-11-26 at 11 50 32
2. Getting this error o trying to add the https://github.com/ubiquity-os-marketplace/daemon-pricing plugin:
Screenshot 2024-11-26 at 11 51 11

Pls fix

**Comment by Keyrxng** [comment]
Both of those plugins fixed: `pricing` was improperly setting of null/undefined vals and `conversation-rewards` was due to the handling of the huge config.

I've just JSON.stringified the huge props for now but I think you are right in that there is no clear clean way of displaying it.

It's messy when you are having to work with `undefined` and `null` values because they should represent a clear fact like `null === they need it but don't have it` and `undefined === can be omitted` but they are used interchangeably

Manually expanded:
image

QA of both plugins:
https://github.com/user-attachments/assets/373142d5-76db-4556-81cd-323df09faf5c

**Comment by Keyrxng** [comment]
> Of course the ideal is to show precisely the ones he can edit. If that can't be done it's better to show all admin orgs than to miss some.

`rest.orgs.listForUser():`
> List [public organization memberships](vscode-file://vscode-app/c:/Users/keyrx/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) for the specified user.
> This method only lists public memberships, regardless of authentication. If you need to fetch all of the organization memberships (public and private) for the authenticated user, use the [List organizations for the authenticated user](vscode-file://vscode-app/c:/Users/keyrx/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) API instead.

This returns all 4 orgs for me and @gentlementlegen, but returns 0 for @zugdev due to your org membership not being public. So we cannot rely on the public `listForUser` endpoints if it does not capture private memberships.

`rest.orgs.listMembershipsForAuthenticatedUser()`:
I set full access with: `scopes: "user repo org"` and it _still_ only returns my org `ubq-testing`. I do have write permissions in `ubiquity` still so it should be appearing for me but isn't.

I think org display should be handled in #26 rather than hold this back but appreciate the clarification for you both, should display all orgs with `write` permissions.

**Comment by Keyrxng** [comment]
This error used to occur on `work.ubq.fi` to me often although it doesn't happen on there anymore, I'm not sure if it was resolved specifically or if it's because of something else, @zugdev do you have any knowledge of this happening on `work.ubq` or was this before your time?

image


---

This is the code that's throwing for me. I assume it's throwing because I'm being specific in the method and when the list methods are used those that would throw are just omitted.

```ts
  public async getGitHubUserOrgs(): Promise<{ org: string; permissions: Record<string, unknown> | undefined, role: string }[]> {
    const octokit = await this.getOctokit();
    const response = await octokit.rest.orgs.listForAuthenticatedUser();
    const authedUser = await octokit.rest.users.getAuthenticated();
    const orgPermission = await octokit.rest.orgs.listMembershipsForAuthenticatedUser();
    const orgs = await octokit.rest.orgs.listForUser({ username: authedUser.data.login });

    const userOrgs = orgPermission.data.map((org) => {
      return {
        org: org.organization.login,
        permissions: org.permissions,
        role: org.role,
      };
    })

    const publicOrgs = []

    for (const org of orgs.data) {
      const orgUserDetails = await octokit.rest.orgs.getMembershipForAuthenticatedUser({
        org: org.login,
      });

      publicOrgs.push({
        org: org.login,
        permissions: undefined,
        role: undefined,
      })
    }

    console.log("User orgs", userOrgs);

    return response.data;
  }
```

**Comment by Keyrxng** [comment]
Considering the above two comments, I'm re-opening this for review and the org selection discussion can be moved into #26

**Comment by zugdev** [comment]
> This error used to occur on `work.ubq.fi` to me often although it doesn't happen on there anymore, I'm not sure if it was resolved specifically or if it's because of something else, @zugdev do you have any knowledge of this happening on `work.ubq` or was this before your time?
> 
> image

This is oAuth rate limit.

**Comment by zugdev** [comment]
Appears like you have solved the problems I found. The only thing I managed to find now is the overlap between table and header:

image

We could create a issue later for me to fix these kind of CSS problems. By taking a very quick look:

1. Nav should be inside of header
2. The height for content does not need to be calculated just use display block use the available height
3. Fix overlapping
4. Make responsive

**Review by zugdev — APPROVED** [review]

**Comment by Keyrxng** [comment]
> We could create a issue later for me to fix these kind of CSS problems. By taking a very quick look:

I'll catch what I can during reviews in the remaining three PRs but yeah that would be good mate cheers.

**Review by rndquu — CHANGES_REQUESTED** [review]
@Keyrxng Check this screenshot:
Screenshot 2024-11-27 at 18 46 00

The "Assistive pricing" plugin is disabled although its config is present in the `.ubiquity-os.config.yml` file. The expected behavior for the "Assistive pricing" is to be in the enabled mode.

Pls fix

**Comment by Keyrxng** [comment]
> @Keyrxng Check this screenshot: 
> 
> The "Assistive pricing" plugin is disabled although its config is present in the `.ubiquity-os.config.yml` file. The expected behavior for the "Assistive pricing" is to be in the enabled mode.
> 
> Pls fix

My apologies @rndquu, fixed now

**Review by rndquu — CHANGES_REQUESTED** [review]
@Keyrxng 

1. For some of the plugins (`User activity watcher`, `Automated merging`, `Telegram kernel`) getting the `No plugin URL found for user-activity-watcher` error:
Screenshot 2024-11-28 at 10 24 57

As far as I understand those plugins work as github actions so it's not clear why we can't set them up.

2. The `command-ask` plugin shows empty UI:
Screenshot 2024-11-28 at 10 29 42

Pls fix

**Comment by Keyrxng** [comment]
Sorted @rndquu

1. PRs opened in relevant repos: related to #27
2. I'm not able to reproduce this. Was it from a fresh cache fetch or did you have it installed?
image

**Comment by Keyrxng** [comment]
I merged https://github.com/ubiquity-os-marketplace/ubiquity-os-kernel-telegram/pull/40 and `kernel telegram` is now working so the other PRs need merged and we should be good to merge this?

image


1. Technically `kernel-telegram` is not built to support being installed by partners at the moment so should probs not be displayed.
2. It has two config entries: A) I detect url(s), if plural, create an entry for each B) See links below

- https://github.com/ubiquity-os/plugin-template/issues/33#issuecomment-2500425415
- https://github.com/ubiquity-os-marketplace/ubiquity-os-kernel-telegram/issues/20

**Review by rndquu — CHANGES_REQUESTED** [review]
Check this screenshot:
Screenshot 2024-11-28 at 19 41 55

Although the `registerWalletWithVerification` is set to `true` in the `yml` config the option in the html UI is not check while the expected behavior is to load current config values.

Pls fix.

**Comment by rndquu** [comment]
> Sorted @rndquu
> 
> 1. PRs opened in relevant repos: related to [`manifest.name` mismatch #27](https://github.com/ubiquity-os/ubiquity-os-plugin-installer/issues/27)
> 2. I'm not able to reproduce this. Was it from a fresh cache fetch or did you have it installed?
>    image

1. So, as far as I understand, when https://github.com/ubiquity-os/ubiquity-os-plugin-installer/issues/24 is solved all plugins setup will be fine because of the correct `name` property.
2. I cleared the cache and error is gone

**Comment by Keyrxng** [comment]
Fixed, I had forgot to set `checked` relevant to the input value.

image


> So, as far as I understand, when https://github.com/ubiquity-os/ubiquity-os-plugin-installer/issues/24 is solved all plugins setup will be fine because of the correct name property.
I cleared the cache and error is gone

Yeah as can be seen with `kernel-telegram` after I merged. Excellent re: cache issue, I tried to improve the fetching of the readme because some would `404` maybe the manifest fetching needs made more robust if we see anything like that again

**Review by rndquu — APPROVED** [review]

**Comment by rndquu** [comment]
> Appears like you have solved the problems I found. The only thing I managed to find now is the overlap between table and header:
> 
> image
> 
> We could create a issue later for me to fix these kind of CSS problems. By taking a very quick look:
> 
> 1. Nav should be inside of header
> 2. The height for content does not need to be calculated just use display block use the available height
> 3. Fix overlapping
> 4. Make responsive

I suppose when all of the critical issues are ready (https://github.com/ubiquity-os/ubiquity-os-plugin-installer/issues/24, https://github.com/ubiquity-os/ubiquity-os-plugin-installer/issues/22, https://github.com/ubiquity-os/ubiquity-os-plugin-installer/issues/20) we can get back to neaten up the css.

**Comment by rndquu** [comment]
@Keyrxng Now you could merge the latest `main` branch into your other PRs from https://github.com/ubiquity-os/ubiquity-os-plugin-installer/pulls

**Pull request merged**
